Add copy-paste to MultipleInputFields

JulianRoesner opened the pull request:
# Short Description
We want that copy & paste also works for MultipleInputFields. This will add this and allow you to use copy & paste on any subfield and it will then correctly paste the data into the field. You can find the relevant ticket [here](https://steuerlotse.atlassian.net/browse/STL-1264?atlOrigin=eyJpIjoiZDIwZmFiZmNhMjA1NDhiMmFjMWE1YTkyNGVmMDhhOTIiLCJwIjoiaiJ9).

# Changes
- Add Javascript to split the pasted data in case of a copy & paste event and set it correctly into the subfields
- Set the input_length as a data field
- Use render_template to render the MultipleInputField

# Feedback
- Does this make sense to you?

frederike-ramin commented on webapp/app/forms/fields.py:
```diff
@@ -96,7 +96,10 @@ def __call__(self, field, **kwargs):
                 joined_input_fields += Markup(self.sub_field_separator)
 
-        return Markup(joined_input_fields)
+        return Markup(render_template('fields/multiple_input_field.html',
+                                      joined_input_fields=joined_input_fields,
+                                      name=field.id,
+                                      input_field_lengths=self.input_field_lengths))
```
Why do you add `name` and `input_field_lengths` as parameters? I don't think you're using them in the template?

JulianRoesner replied:
Good catch, input_field_length is now set as an additional attribute of the subfield and can go. `name` is used in line 6 in the template: `const sub_fields = document.querySelectorAll('input[name={{ name }}]');`

frederike-ramin commented on webapp/app/templates/fields/multiple_input_field.html:
```diff
@@ -0,0 +1,24 @@
+        sub_fields.forEach(input => {
+            input.addEventListener('paste', e => {
+                const data = e.clipboardData.items[0];
+                if(data.kind == 'string' && data.type == 'text/plain') {
+                    data.getAsString(str => {
+                        let start_idx = 0
+                        sub_fields.forEach(sub_field => {
+                            const length = parseInt(sub_field.dataset.fieldLength);
+                            const paste_content = str.substr(start_idx, length);
+                            sub_field.value = paste_content || '';
+                            start_idx += length;
+                        });
+                    });
+                }
```
I don't think this works if we have spaces in the copied data? So I think we should add an additional case for this.

JulianRoesner replied:
I think I do not understand what you mean. For me, this works also with spaces in the copied data.

JulianRoesner replied:
To make sure that that the data is copied in the correct format (e.g. 24.12.1964 -> 24 12 1964 for the three fields), we will check the pattern and change the copied data correctly.

JulianRoesner replied:
I've changed it a little and we are now masking the paste content before we paste it into the corresponding fields.

frederike-ramin reviewed (changes requested):
The code looks good, generally. However, the functionality didn't work at all for me, neither on Chrome nor Firefox (my Safari is not reacting at all, so I can't try it there -.-)
Instead, no content is added, not even the first x chars as before. It appears and then directly disappears again.

JulianRoesner commented:
I've tested it again with Firefox and Chrome and it works for me. So maybe let's just talk and go over it together?

JulianRoesner commented:
> However, the functionality didn't work at all for me, neither on Chrome nor Firefox (my Safari is not reacting at all, so I can't try it there -.-)
It worked locally but not for you because adding the `data-field-length` has not been pushed yet 🙈

JulianRoesner commented:
@nfelger Could you have another look at this and maybe check whether it works for you on Chrome and also have a look if masking the `paste_content` makes sense to you?

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                 const sub_fields = document.querySelectorAll('input[name="unlock_code"]');
+                 sub_fields.forEach(input => {
+                     input.addEventListener('input', e => {
+                         console.log("old value" + window.memo_values[input.id]);
+                         console.log("Hallo" + input.value);
+                         const data = input.value;
+                         console.log(input.dataset.fieldLength);
+                         console.log(data.length, Number(input.dataset.fieldLength));
+                         console.log(data.length > Number(input.dataset.fieldLength));
+                         let field_to_focus = input
+                         if (data.length > Number(input.dataset.fieldLength)) {
+
+                             console.log("start splitting");
+                             let start_idx = 0;
+                             let start = false;
+                             sub_fields.forEach(sub_field => {
+                                 if (start || sub_field == input){
+                                     start = true;
+                                     const length = parseInt(sub_field.dataset.fieldLength);
+                                     const paste_content = data.substr(start_idx, length);
+                                     console.log("split content", paste_content, length);
+                                     if (paste_content.length > 0){
+                                        field_to_focus = sub_field;
+                                     };
+                                     sub_field.value = paste_content || '';
+                                     start_idx += length;
+                                 }
+                             });
+                             console.log(field_to_focus);
```
Please remove the console.log statements, they should not go into production.

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+            $(document).ready(function(){
+                $('.date_input').mask('00.00.0000');
+                $('input[name="unlock_code"]').mask('A', {
+                    translation: {
+                        "A": { pattern: /[\w]/, recursive: true }
+                    }
+                });
+                window.memo_values  = new Array();
+
+                 const sub_fields = document.querySelectorAll('input[name="unlock_code"]');
+                 sub_fields.forEach(input => {
+                     input.addEventListener('input', e => {
+                         console.log("old value" + window.memo_values[input.id]);
+                         console.log("Hallo" + input.value);
+                         const data = input.value;
+                         console.log(input.dataset.fieldLength);
+                         console.log(data.length, Number(input.dataset.fieldLength));
+                         console.log(data.length > Number(input.dataset.fieldLength));
+                         let field_to_focus = input
+                         if (data.length > Number(input.dataset.fieldLength)) {
+
+                             console.log("start splitting");
+                             let start_idx = 0;
+                             let start = false;
+                             sub_fields.forEach(sub_field => {
+                                 if (start || sub_field == input){
+                                     start = true;
+                                     const length = parseInt(sub_field.dataset.fieldLength);
+                                     const paste_content = data.substr(start_idx, length);
+                                     console.log("split content", paste_content, length);
+                                     if (paste_content.length > 0){
+                                        field_to_focus = sub_field;
+                                     };
+                                     sub_field.value = paste_content || '';
+                                     start_idx += length;
+                                 }
+                             });
+                             console.log(field_to_focus);
+                             field_to_focus.focus();
+                         }
```
Is there a more specific place for this than `base_step_form`? There's no need to include this on every step form. (But maybe we don't have a template for unlock_code fields?)

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                     input.addEventListener('input', e => {
+                         console.log("old value" + window.memo_values[input.id]);
+                         console.log("Hallo" + input.value);
+                         const data = input.value;
```
You can use `e.target` instead of `input` here. It's a bit cleaner because it doesn't rely on the reference from the closure and makes it easier to extract the function later if we need to.

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                             let start_idx = 0;
+                             let start = false;
+                             sub_fields.forEach(sub_field => {
+                                 if (start || sub_field == input){
```
Makes no difference here, but please always use `===` instead of `==` in JS.

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                             sub_fields.forEach(sub_field => {
+                                 if (start || sub_field == input){
+                                     start = true;
+                                     const length = parseInt(sub_field.dataset.fieldLength);
```
Using `parseInt` here and `Number` on line 62. We should be consistent. `parseInt` makes more sense to me, but open to discussing it.

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                                 if (start || sub_field == input){
+                                     start = true;
+                                     const length = parseInt(sub_field.dataset.fieldLength);
+                                     const paste_content = data.substr(start_idx, length);
```
Re: naming - `paste_content` isn't just pasted content, right? It's any entered data. If so, maybe call it `unlock_code_fragment` or something like that?

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                        "A": { pattern: /[\w]/, recursive: true }
+                    }
+                });
+                window.memo_values  = new Array();
```
This isn't being used.

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/basis/base_step_form.html:
```diff
@@ -39,4 +39,50 @@
+                });
+                window.memo_values  = new Array();
+
+                 const sub_fields = document.querySelectorAll('input[name="unlock_code"]');
```
I'm a bit confused here - is this bit only for unlock_code fields, not all multiple input fields? Is it to support autocomplete?

JulianRoesner replied:
This script was added by mistake (s. below)

nfelger commented on webapp/app/templates/fields/multiple_input_field.html:
```diff
@@ -0,0 +1,25 @@
+                    data.getAsString(str => {
+                        let start_idx = 0
+                        sub_fields.forEach(sub_field => {
+                            masked_str = $(sub_field).masked(str);
```
I like it. Why not!

JulianRoesner replied:
👍

JulianRoesner replied:
👍

nfelger reviewed (changes requested):
Works for me as expected in Chrome (well, Brave). With spaces and without.

I'm confused about the part in base_step_form.html and have left a bunch of comments / questions there.

Let's jump on a huddle to discuss if you want.

JulianRoesner commented on webapp/app/templates/fields/multiple_input_field.html:
```diff
@@ -11,9 +11,14 @@
                         let start_idx = 0
                         sub_fields.forEach(sub_field => {
-                            masked_str = $(sub_field).masked(str);
+                            try{
+                                masked_str = $(sub_field).masked(str);
+                            }catch (TypeError) {
+                                masked_str = str;
+                            }
```
I've added this because it won't work on multiple Input fields without a `data-mask` attribute otherwise.

JulianRoesner commented:
> I'm confused about the part in base_step_form.html and have left a bunch of comments / questions there.

You have every right to be confused. Unfortunately, I've overlooked that the autocomplete experiment code got into the last commit. So everything in the `base_step_form.html ` is not for the copy-paste feature and I've deleted it now. I am very sorry.

While testing it, I've noticed that it does not work for the unlock field because it has no `data-mask` attribute. So I put the mask function into a try-catch block. I've also added a `data-mask` attribute to the unlock code field in #119 and I've tested it with Chrome, Safari and Firefox -> works fine

nfelger reviewed (approved):
Got it, thanks! Looks good to go now!
